### ISSUE TITLE
A is not >=0 when using `train_unsupervised`

### ISSUE THREAD

**Issue by EArtiges** [jvendrow/NeuralNMF#2]
Hi! 

I am trying to use this library for a problem very similar to the NewsGroup example you provide. When I run the example notebook on my laptop (after setting `epoch=10` to save time, granted), I get values inferior to zero for `A.min()`.

Looking at the code more in detail, in `train.py`, line 359 in `train_supervised`, `A.data = A.data.clamp(min = 0)`, is not replicated in `train_supervised` (it should sit on line 210 I believe?), which could be the reason why.

Could you look into this and let me know? Best!

**Comment by Prunoideae** [comment]
There are actually quite a lot of problems with this model. The training is sensitive to the means of data, and you need to do a lot of parameter tuning to get a good-looking result, but in the end, it's not good-looking either.

Since the model targets at decomposing A1, S1, A2, S2... Those are actually a re-decomposition of the W/H matrices outputted by the last layer of decomposition. I suggest you simply chain NMF decompositions to get better (and numerically more stable) results:

```python

k1, k2 = 12, 5

# ---- Load and preprocess data ----
# ...
# ----------------------------------

# ---- NMF (pass 1) ----

W1, H1, n_iter = non_negative_factorization(
    X,
    n_components=k1,
    init="nndsvda",
    random_state=42,
    max_iter=2000,
    beta_loss='kullback-leibler',
    solver='mu'
)

print(W1.shape, H1.shape)  # (obs, k1), (k1, var)
# ----------------------

# ---- NMF (pass 2) ----

W2, H2, n_iter = non_negative_factorization(
    H1,
    n_components=k2,
    init="nndsvda",
    random_state=42,
    max_iter=2000,
    # Since it's not counts anymore, we use Frobenius
)

print(W2.shape, H2.shape)  # (k1, k2), (k2, var)
print(((W1 @ W2) @ H2).shape)  # (obs, var)
# ----------------------

# ---- Evaluation ----
X_approx_1 = W1 @ H1
X_approx_2 = (W1 @ W2) @ H2
# ...
# ----------------------
```

Which will produce the A1...An and H1...Hn matrices you need.

By the way, did you find better tools for hierarchical NMF? It would be great to know if there are other available tools to use.

**Comment by EArtiges** [comment]
Interesting. Can you develop on the training sensitivity to the mean?

If I'm not mistaken, your solution is what I was doing before I found this model and the associated paper: you're doing a first low level factorization (high k), then factorize that to get the higher levels factors (low k). But nothing tells us that the resulting `(W_1, H_1, W_2, H_2` minimize any kind of error (i.e it does not optimize the whole decomposition end to end, it only optimizes each step separately) which is meant to be the main selling point of the method in this repo.

I haven't found any other solution so far, but the idea developped here is extremely elegant and I think it's worth investigating. My main issue with it is that it's pretty slow- maybe someone more motivated than me could take that idea and re-write it in C or something (maybe Rust? I don't know if the NMF can be easily parallelized).


As an alternative to this, there might be a way to plug the high-k factorization into the low-k factorization and analytically derive an algorithm to perform `n` decompositions at the same time without going down the Neural Net route- after all, it's a gradient descent- but then again, I am not interested in this enough to have investigated it myself.

**Comment by Prunoideae** [comment]
> Interesting. Can you develop on the training sensitivity to the mean?

After looking into the code, I think the actual cause might be the sum of the features. Each of my samples sums to 1e6, which caused the loss to skyrocket, and the reconstructed matrix to perform poorly in evaluations. When the sum reaches 1e4, the loss curve improves, and the results are better, but it's still far from the 1-step decomposition using different k values directly. I think it's because the `Fro_Norm` uses `nn.MSELoss` without normalization, so the input sum actually matters, but I wonder why it does not help when the classification type is `CrossEntropyLoss` instead of `L2`, maybe there're something in the training/model code too.

> As an alternative to this, there might be a way to plug the high-k factorization into the low-k factorization and analytically derive an algorithm to perform `n` decompositions at the same time

I think it is possible to implement an NMF with a custom `mu` or `cd` update method that fuses two decompositions and change the objective function to `L(W1, H1, W2, H2) = beta_divergence(X, W1H1) + beta_divergence(X, W1W2H2) + regularization`. 

But the accuracy of performing two separate decompositions is quite acceptable based on my evaluation. It's probably because the `high k -> low k` decomposition tries to extract more general structures in the data. And, transitioning to `low k` only destroyed the detailed, local patterns, while the general patterns were not quite affected. So, it's good enough for my case.

**Comment by Prunoideae** [comment]
Oh great, I got a neural version of it with the help of AI:

```python
import torch
import torch.nn as nn
import torch.nn.functional as F


class hNMF(nn.Module):
    def __init__(self, W1, H1, W2, H2, lr=1e-3, eps=1e-8):
        """
        Initialize hNMF with individual decompositions.
        """
        super(hNMF, self).__init__()
        self.eps = eps

        # Initialize parameters as learnable
        self.W1 = nn.Parameter(torch.FloatTensor(W1))
        self.H1 = nn.Parameter(torch.FloatTensor(H1))
        self.W2 = nn.Parameter(torch.FloatTensor(W2))
        self.H2 = nn.Parameter(torch.FloatTensor(H2))

        # Optimizer for all parameters
        self.optimizer = torch.optim.Adam(self.parameters(), lr=lr)

    def forward(self):
        # Ensure non-negativity
        W1 = torch.clamp(self.W1, min=0)
        H1 = torch.clamp(self.H1, min=0)
        W2 = torch.clamp(self.W2, min=0)
        H2 = torch.clamp(self.H2, min=0)

        # Pass 1 reconstruction: W1 @ H1
        recon_pass1 = torch.matmul(W1, H1)

        # Pass 2 reconstruction: (W1 @ W2) @ H2
        recon_pass2 = torch.matmul(W1, torch.matmul(W2, H2))

        return recon_pass2, recon_pass1

    def kl_divergence(self, X, X_recon):
        """KL divergence for count data"""
        X_recon = X_recon + self.eps
        return torch.sum(X * torch.log((X + self.eps) / X_recon) - X + X_recon)

    def loss_fn(self, X, recon, alpha=0.5, loss_type="kl"):
        """
        Parameters:
        -----------
        X : torch.Tensor
            Input data
        recon : tuple
            (recon_pass2, recon_pass1)
        alpha : float
            Weight for pass1 loss (default: 0.5)
        loss_type : str
            'kl' for KL divergence, 'mse' for mean squared error
        """
        recon_pass2, recon_pass1 = recon

        if loss_type == "kl":
            loss_pass2 = self.kl_divergence(X, recon_pass2)
            loss_pass1 = self.kl_divergence(X, recon_pass1)
        else:  # MSE
            loss_pass2 = F.mse_loss(recon_pass2, X, reduction="sum")
            loss_pass1 = F.mse_loss(recon_pass1, X, reduction="sum")

        total_loss = loss_pass2 + alpha * loss_pass1
        return total_loss, loss_pass2, loss_pass1

    def fit(self, X, n_epochs=100, alpha=0.5, loss_type="kl", verbose=True):
        """
        Train the hierarchical NMF model.

        Parameters:
        -----------
        X : numpy array or torch tensor
            Input data matrix
        n_epochs : int
            Number of training epochs
        alpha : float
            Weight for pass1 loss relative to pass2
        loss_type : str
            'kl' for count data, 'mse' for continuous data
        verbose : bool
            Print progress every 10 epochs
        """
        # Convert to tensor if needed
        if not isinstance(X, torch.Tensor):
            X_tensor = torch.FloatTensor(X)
        else:
            X_tensor = X

        losses = {"total": [], "pass2": [], "pass1": []}

        for epoch in range(n_epochs):
            self.optimizer.zero_grad()

            # Forward pass
            recon = self.forward()

            # Compute loss
            total_loss, loss_pass2, loss_pass1 = self.loss_fn(
                X_tensor, recon, alpha, loss_type
            )

            # Backward pass
            total_loss.backward()

            # Update parameters
            self.optimizer.step()

            # Enforce non-negativity constraints
            with torch.no_grad():
                self.W1.clamp_(min=0)
                self.H1.clamp_(min=0)
                self.W2.clamp_(min=0)
                self.H2.clamp_(min=0)

            # Track losses
            losses["total"].append(total_loss.item())
            losses["pass2"].append(loss_pass2.item())
            losses["pass1"].append(loss_pass1.item())

            if verbose and (epoch % 10 == 0 or epoch == n_epochs - 1):
                print(
                    f"Epoch {epoch:4d}/{n_epochs}: "
                    f"Total = {total_loss.item():.2f}, "
                    f"Pass2 = {loss_pass2.item():.2f}, "
                    f"Pass1 = {loss_pass1.item():.2f}"
                )

        return losses

    def get_components(self):
        """Return learned components as numpy arrays"""
        with torch.no_grad():
            return (
                torch.clamp(self.W1, min=0).cpu().numpy(),
                torch.clamp(self.H1, min=0).cpu().numpy(),
                torch.clamp(self.W2, min=0).cpu().numpy(),
                torch.clamp(self.H2, min=0).cpu().numpy(),
            )

    def transform(self):
        """Get final reconstructions"""
        with torch.no_grad():
            recon_pass2, recon_pass1 = self.forward()
            return recon_pass2.cpu().numpy(), recon_pass1.cpu().numpy()
```

Usage:
```python
model = hNMF(W1, H1, W2, H2, lr=1e-3)
losses = model.fit(X, n_epochs=1000, alpha=1, loss_type="kl", verbose=True)
W1_joint, H1_joint, W2_joint, H2_joint = model.get_components()
X_hNMF_2, X_hNMF_1 = model.transform()
```

My evaluation metrics (the lower the better). It shows that the hNMF model tried to improve the pass 2 result with minor compensation in pass 1.
```
Pass 1 BC distance: mean=0.18754090435953819, std=0.056418287874812956, min=0.08497067434294654, max=0.3794236075907914
Pass 2 BC distance: mean=0.2997175091944561, std=0.10121599746211396, min=0.10663730977830578, max=0.530428396562451
hNMF Pass 1 BC distance: mean=0.18907519478852727, std=0.05651524144086005, min=0.0831323788150552, max=0.3719962918326361
hNMF Pass 2 BC distance: mean=0.28775443738906303, std=0.09652222401857381, min=0.09761018982497902, max=0.49461597517418654
```
